gpii/node_modules/lifecycleManager/src/UserLogonRequest.js

```diff
@@ -0,0 +1,228 @@
+            args: ["{that}", "{lifecycleManager}", "{arguments}.0"]
+                                                   // logonState
+        }
+    },
+    listeners: {
+        "onSuccess.handleSuccessRequest": {
+            listener: "gpii.lifecycleManager.userLogonRequest.handleSuccessRequest",
+            args: ["{that}", "{arguments}.0"]
+        },
+        "onError.handleFailedRequest": {
+            listener: "gpii.lifecycleManager.userLogonRequest.handleFailedRequest",
+            args: ["{that}", "{arguments}.0"],
+            priority: "after:handleErrors"
+        }
+    }
+});
+
+gpii.lifecycleManager.userLogonRequest.handleRequest = function (that, lifecycleManager, logonState) {
+    var gpiiKey = that.options.gpiiKey;
+    var activeGpiiKeys = lifecycleManager.getActiveSessionGpiiKeys(); // find currently logged in user:
+
+    // keep track of the last non-noUser key for /proximityTriggered endpoint to verify its debounce rule
+    // when the the current active key is "noUser".
+    if (gpiiKey !== "noUser") {
+        gpii.lifecycleManager.userLogonHandling.updateLastNonNoUserLogon(lifecycleManager, gpiiKey, logonState);
+    }
+
+    // handle reset requests
+    if (gpiiKey === "reset") {
+        that.handleResetGpiiKey();
+    }
+
+    // handle keyin requests
+    if (logonState === "login") {
+        // if another user, excluding "noUser", is already logged in
+        if (activeGpiiKeys.length === 1 && activeGpiiKeys[0] !== "noUser") {
+            that.errorResponse("Got log in request from user " + gpiiKey + ", but the user " +
+                activeGpiiKeys[0] + " is already logged in. So ignoring login request.", 409);
+        } else {
+            that.loginUser(gpiiKey);
+        }
+    }
+
+    // handle keyout requests
+    if (logonState === "logout") {
+        if (activeGpiiKeys[0] !== gpiiKey) {
+            that.errorResponse("Got logout request from user " + gpiiKey + ", but the user " +
+                activeGpiiKeys[0] + " is logged in. So ignoring the request.", 409);
+        } else {
+            that.logoutUser(gpiiKey);
+        }
+    }
+
+    // handle getGpiiKey requests
+    if (logonState === "getGpiiKey") {
+        that.withSession(function (session, gpiiKeys) {
+            fluid.log("Reporting currently logged in GPII key(s) as: ", gpiiKeys);
+            that.events.onSuccess.fire(gpiiKeys);
+        }, "No user currently logged in to the system");
+    }
+};
+
+gpii.lifecycleManager.userLogonRequest.handleSuccessRequest = function (that, payload) {
+    that.options.requestPromise.resolve(payload);
+    that.destroy();
+};
+
+gpii.lifecycleManager.userLogonRequest.handleFailedRequest = function (that, err) {
+    that.options.requestPromise.reject(err);
+    that.destroy();
+};
+
+// Define the grade to be used for key-in and key-out
+gpii.lifecycleManager.userLogonRequest.handlerGrades = fluid.freezeRecursive({
```

Comment:
No great reason to freeze this - one of the principal values of having it in a public structure is that someone might conceivably want to patch these